src/libopensc/pkcs15-skey.c

```diff
@@ -140,13 +143,36 @@ sc_pkcs15_decode_skdf_entry(struct sc_pkcs15_card *p15card, struct sc_pkcs15_obj
 
         /* Fill in defaults */
 	memset(&info, 0, sizeof(info));
+	info.native = 1;
 
 	r = sc_asn1_decode(ctx, asn1_skey, *buf, *buflen, buf, buflen);
 	if (r == SC_ERROR_ASN1_END_OF_CONTENTS)
 		return r;
 	LOG_TEST_RET(ctx, r, "ASN.1 decoding failed");
-	if (asn1_skey_choice[0].flags & SC_ASN1_PRESENT)
+	if (asn1_skey_choice[0].flags & SC_ASN1_PRESENT) {
 		obj->type = SC_PKCS15_TYPE_SKEY_GENERIC;
+
+		/* info.key_type = CKK_AES;*/	/* defaults to CKK_AES. TODO: implement choice 15 and GenericKeyAttributes like defined in ISO7816-15 and set key type accordingly. */
+
+	/* Check key type. framework-pkcs15 recognizes one type per key, and AES is the only algorithm supported for
+	 * SKEY_GENERIC type keys, so just check if this key is AES compatible. */
+
+		for (i = 0; i < SC_MAX_SUPPORTED_ALGORITHMS && info.algo_refs[i] != 0; i++) {
+			for (ii = 0; ii < SC_MAX_SUPPORTED_ALGORITHMS && p15card->tokeninfo != 0; ii++) {
+				if (info.algo_refs[i] == p15card->tokeninfo->supported_algos[ii].reference) {
+				    temp_oid = p15card->tokeninfo->supported_algos[ii].algo_id;
+				    temp_oid.value[8] = -1; /* strip off AES subtype octet*/
+
+				    if (sc_compare_oid(&id_aes, &temp_oid))
+					if (info.key_type == 0)	{
+					    info.key_type = CKK_AES;
```

Comment:
Is PKCS#15/ISO 7816-15 really mixing some context from PKCS#11 into this?

Reply:
No. PKCS#15 has OIDs of supported algorithms listed in tokenInfo. Here we check, if the algorithm referenced in the PKCS#15 SecretKeyObject matches with AES algorithm's OID and set the key type to CKK_AES for OpenSC's internal use, out of PKCS#15's context.

I did ponder for a while, should I include the pkcs#11 header here or define new algorithm types in the PKCS#15 headers. I ended up doing it this way, because CKK_ key type definitions are already used in skey_pkcs15_skey_info before my pull request. Please take a look at line 2315 of framework-pkcs15.c in current master, function pkcs15_create_secret_key: 
	    skey_info->key_type = key_type; /* PKCS#11 CKK_* */

I decided to use the CKK_AES defintion in pkcs15-skey to be consistent with this. At least we have to use the same set of definitions and I don't know how much code it would affect to change it now.